return HTTP response for error status as part of exception

SteveL-MSFT opened the pull request:
Addresses #2193 

When Invoke-WebRequest or Invoke-RestMethod receives a HTTS status error code, an exception is returned and the user can't get the HTTP response without additional work.  This change makes it more consistent with Windows PowerShell in that a Response property which contains the HttpResponse is a member of the resulting Exception.

SteveL-MSFT committed:
added http response to new HttpResponseException to provide similar experience to Windows PowerShell
added tests for invoke-webrequest and invoke-restmethod for http error cases

SteveL-MSFT committed:
added http response to new HttpResponseException to provide similar experience to Windows PowerShell
added tests for invoke-webrequest and invoke-restmethod for http error cases

fixed bad merge

SteveL-MSFT committed:
fixed bad merge

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,50 @@
     /// <summary>
+    /// Exception class for webcmdlets to enable returning HTTP error response
+    /// </summary>    
+    public class HttpResponseException : Exception
```
Should this class be sealed?  Or is there a reason to derive from it?

SteveL-MSFT replied:
Agree, no reason to derive from it

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,50 @@
+        /// <summary>
+        /// HTTP error status code
+        /// </summary>
+        public HttpStatusCode Status
```
It looks like this property is never set in the error handling code below.  Does this just return the response.StatusCode?

SteveL-MSFT replied:
Was originally adding the individual properties before I saw the example of Windows PowerShell just having a Response property.  Didn't remove them.  Will remove.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,50 @@
+        /// <summary>
+        /// HTTP error headers
+        /// </summary>
+        public HttpResponseHeaders Headers
```
Same for this property.  Is this _response.Headers?

SteveL-MSFT replied:
Will remove.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,50 @@
+        /// <summary>
+        /// HTTP error response
+        /// </summary>
+        public HttpResponseMessage Response
```
I believe we normally use automatic implemented getter/setter property syntax:
```powershell
public HttpResponseMessage Response { get; set; }
```

SteveL-MSFT replied:
Will fix

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +391,24 @@ protected override void ProcessRecord()
+                        {
+                            string message = String.Format(CultureInfo.CurrentCulture, WebCmdletStrings.ResponseStatusCodeFailure,
+                                response.StatusCode.ToString(), response.ReasonPhrase);
+                            HttpResponseException httpEx = new HttpResponseException(message);
```
You could use this syntax here for creating the HttpResponseException:
```
var httpEx = new HttpResponseException(message)
{
    Response = response,
    Status = ...
    Headers = ...
};
```
Or you could make the constructor take all property values and make the properties read only.

SteveL-MSFT replied:
Will change constructor

PaulHigin commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -399,6 +399,18 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+        $result = ExecuteWebCommand -command $command
+
+        $result.Error.Exception | Should BeOfType Microsoft.PowerShell.Commands.HttpResponseException
+        $result.Error.Exception.Response.StatusCode | Should Be 418
```
Should we test for Exception.Headers and Exception.Status too?

SteveL-MSFT replied:
Those are getting removed

SteveL-MSFT committed:
sealed HttpResponseException class, used auto-accessors, removed unused Headers and Status members, added response to constuctor based on feedback

SteveL-MSFT committed:
Merge branch 'weberror'

SteveL-MSFT commented:
@PaulHigin thanks for the feedback, addressed each of them

PaulHigin commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,27 @@
+        /// <summary>
+        /// HTTP error response
+        /// </summary>
+        public HttpResponseMessage Response { get; set; }
```
We should make the setter private { get; private set; }.  Otherwise LGTM.

daxian-dbw replied:
I agree. `WebException.Response` only has a getter.

SteveL-MSFT replied:
Makes sense

daxian-dbw commented:
@SteveL-MSFT We have another similar PR here #3089, please take a look at the comments in that PR to see any of them apply in this fix as well.
I suggest we have our `HttpResponseException` derives from `System.Net.Http.HttpRequestException`, so that `catch [HttpRequestException] { }` would continue to work after this change.

daxian-dbw commented:
@2xmax (the author of #3089). would you mind take a look?

2xmax commented:
will the new version of the cmdlet output the content if the code was unsuccessful? I could say in the context of my scope, that it was a real issue and we even already rewrote our code to curl instead of powershell whereas the status code was not a case, afaik we were able to parse it using regex:)).

daxian-dbw commented:
@2xmax you are right, this PR only covers partial of #3089. It doesn't capture the content from the response and put it in `ErrorDetail` as full powershell does (see [code here](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/FullClr/WebRequestPSCmdlet.FullClr.cs#L469)).

@SteveL-MSFT, do you want to address it in this same PR or a different one?

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,27 @@
     /// <summary>
+    /// Exception class for webcmdlets to enable returning HTTP error response
+    /// </summary>    
+    public sealed class HttpResponseException : Exception
```
I suggest we make `HttpResponseException` derive from `HttpRequestException`, so that `{ Invoke-WebRequest <url> } catch [HttpRequestException] { }` will continue to work after this change.

SteveL-MSFT replied:
Agree

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,23 @@ protected override void ProcessRecord()
+                            HttpResponseException httpEx = new HttpResponseException(message, response);
+                            ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
+                            er.ErrorDetails = new ErrorDetails(message);
+                            ThrowTerminatingError(er);
```
The message for `ErrorDetails` should be the content of the response (after using regex replace to remove tags). This was what #3089 tried to fix, and the corresponding code in full powershell is [here](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/FullClr/WebRequestPSCmdlet.FullClr.cs#L469)

SteveL-MSFT replied:
Will make the change

2xmax commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -642,6 +654,17 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
         $result = ExecuteWebCommand -command $command
         $result.Error | Should BeNullOrEmpty
     }
+
```
pls add a test with a response content

SteveL-MSFT replied:
will do

SteveL-MSFT committed:
derived HttpResponseException from HttpRequestException so other catches continue to work
added response stripping html tags to ErrorDetails
added test to verify response is in ErrorDetails

SteveL-MSFT commented:
Addressed feedback.  This should also address https://github.com/PowerShell/PowerShell/issues/2113

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,29 @@ protected override void ProcessRecord()
+                            HttpResponseException httpEx = new HttpResponseException(message, response);
+                            ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
+                            string detailMsg = "";
+                            using (StreamReader reader = new StreamReader(StreamHelper.GetResponseStream(response)))
```
Exceptions may be thrown out in `StreamHelper.GetResponseStream(response)` or `reader.ReadToEnd()`. We should put the `using` block in a try/catch-all block, just like [what we did](https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/FullClr/WebRequestPSCmdlet.FullClr.cs#L473) in full ps.

SteveL-MSFT replied:
ok

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,29 @@ protected override void ProcessRecord()
+                        if (!response.IsSuccessStatusCode)
+                        {
+                            string message = String.Format(CultureInfo.CurrentCulture, WebCmdletStrings.ResponseStatusCodeFailure,
+                                Convert.ToInt32(response.StatusCode).ToString(), response.ReasonPhrase);
```
> Convert.ToInt32(response.StatusCode).ToString()

A minor comment: maybe just `(int)response.StatusCode` as in [HttpResponseMessage.cs](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L161)?
If we want to avoid boxing the int argument, then it should be `((int)response.StatusCode).ToString()`, but maybe it's an overkill 😄

SteveL-MSFT replied:
I'm ok with just the cast

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,29 @@ protected override void ProcessRecord()
+                                // remove HTML tags making it easier to read
+                                detailMsg = System.Text.RegularExpressions.Regex.Replace(reader.ReadToEnd(), "<[^>]*>","");
+                            }
+                            er.ErrorDetails = new ErrorDetails(detailMsg);
```
`er.ErrorDetails` should be set only if `detailMsg` is not an empty string..

SteveL-MSFT replied:
ok

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -399,6 +399,19 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+        $result.Error.Exception | Should BeOfType Microsoft.PowerShell.Commands.HttpResponseException
+        $result.Error.Exception.Response.StatusCode | Should Be 418
+        $result.Error.Exception.Response.ReasonPhrase | Should Be "I'm a teapot"
+        $result.Error.Exception.Message | Should Be "Response status code does not indicate success: 418 (I'm a teapot)."
```
Minor comment: this might fail if we are going to run this test on a localized machine.

SteveL-MSFT replied:
@daxian-dbw Is a match good enough for the part returned from the server?

SteveL-MSFT replied:
changed to Match

daxian-dbw replied:
Yes, I think so, `$result.Error.Exception.Message | Should BeLike * 418 (I'm a teapot)` should be good, since they are returned from the Http service, not affected by PS localization.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -642,6 +655,19 @@ Describe "Invoke-RestMethod tests" -Tags "Feature" {
+        $result.Error.Exception | Should BeOfType Microsoft.PowerShell.Commands.HttpResponseException
+        $result.Error.Exception.Response.StatusCode | Should Be 418
+        $result.Error.Exception.Response.ReasonPhrase | Should Be "I'm a teapot"
+        $result.Error.Exception.Message | Should Be "Response status code does not indicate success: 418 (I'm a teapot)."
```
Same minor comment here.

SteveL-MSFT replied:
changed to Match

SteveL-MSFT committed:
added try..catch..finally around reading response stream and checking if detailMsg is empty
changed error message test to be localization friendly by just matching the part returned from server

SteveL-MSFT committed:
added try..catch around reading response stream and checking if detailMsg is empty
changed error message test to be localization friendly by just matching the part returned from server

2xmax commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -20,6 +20,27 @@
     /// <summary>
+    /// Exception class for webcmdlets to enable returning HTTP error response
+    /// </summary>    
+    public sealed class HttpResponseException : HttpRequestException
```
I suggest moving the class into a separate file

SteveL-MSFT replied:
I did a search of the existing source and it appears that in some cases Exceptions are declared in a separate file (like SessionStateExceptions.cs), but in many other cases, the Exception is declared where it's used (like parserutils.cs).

2xmax commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,39 @@ protected override void ProcessRecord()
+                            HttpResponseException httpEx = new HttpResponseException(message, response);
+                            ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
+                            string detailMsg = "";
+                            try
```
I suggest using try-catch-finally to avoid nesting

SteveL-MSFT replied:
ok

daxian-dbw replied:
@2xmax do you mean like this?
```
StreamReader reader = null;
try
{
  StreamReader reader = new StreamReader(StreamHelper.GetResponseStream(response))
  ....
}
catch { }
finally { if (reader != null) { reader.Dispose(); } }
```

SteveL-MSFT committed:
remove accidentally included file
moved allocation of StreamReader into try block

SteveL-MSFT committed:
Merge branch 'master' of github.com:SteveL-MSFT/PowerShell

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs:
```diff
@@ -347,14 +368,41 @@ protected override void ProcessRecord()
+                            HttpResponseException httpEx = new HttpResponseException(message, response);
+                            ErrorRecord er = new ErrorRecord(httpEx, "WebCmdletWebResponseException", ErrorCategory.InvalidOperation, request);
+                            string detailMsg = "";
+                            StreamReader reader = new StreamReader(StreamHelper.GetResponseStream(response));
```
I'm afraid you need to put `new StreamReader(StreamHelper.GetResponseStream(response));` in the try block. Maybe like this:
```
string detailMsg = "";
StreamReader reader = null;
try
{
  reader = new StreamReader(StreamHelper.GetResponseStream(response))
  detailMsg = System.Text.RegularExpressions.Regex.Replace(reader.ReadToEnd(), "<[^>]*>","");
}
catch { }
finally { if (reader != null) { reader.Dispose(); } }
```

SteveL-MSFT replied:
fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs.orig:
```diff
@@ -0,0 +1,610 @@
+﻿#if CORECLR
+
+/********************************************************************++
```
And this file got committed by accident 😄

SteveL-MSFT replied:
fixed

SteveL-MSFT committed:
move allocation of StreamReader into try block

daxian-dbw committed:
Add null check before disposing `reader`
